events/nats.go

```diff
@@ -0,0 +1,377 @@
+	msg.EventType = string(eventType)
+
+	msgb, err := json.Marshal(msg)
+	if err != nil {
+		return err
+	}
+
+	// retry publishing for a while
+	options := []nats.PubOpt{
+		nats.RetryAttempts(-1),
+	}
+
+	subject := fmt.Sprintf("%s.%s.%s", n.parameters.PublisherSubjectPrefix, resType, eventType)
+	if _, err := n.jsctx.PublishAsync(subject, msgb, options...); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Subscribe to all configured SubscribeSubjects
+func (n *NatsJetstream) Subscribe(ctx context.Context) (MsgCh, error) {
+	if n.jsctx == nil {
+		return nil, errors.Wrap(ErrNatsJetstreamAddConsumer, "Jetstream context is not setup")
+	}
+
+	// Subscribe as a pull based subscriber
+	if n.parameters.Consumer != nil && n.parameters.Consumer.Pull {
+		if err := n.subscribeAsPull(ctx); err != nil {
+			return nil, err
+		}
+	}
+
+	// regular Async subscription
+	for _, subject := range n.parameters.SubscribeSubjects {
+		subscription, err := n.jsctx.Subscribe(subject, n.subscriptionCallback, nats.Durable(n.parameters.AppName))
+		if err != nil {
+			return nil, errors.Wrap(ErrSubscription, err.Error()+": "+subject)
+		}
+
+		n.subscriptions = append(n.subscriptions, subscription)
+	}
+
+	return n.subscriberCh, nil
+}
+
+// subscribeAsPull sets up the pull subscription
+func (n *NatsJetstream) subscribeAsPull(ctx context.Context) error {
+	if n.jsctx == nil {
+		return errors.Wrap(ErrNatsJetstreamAddConsumer, "Jetstream context is not setup")
+	}
+
+	for _, subject := range n.parameters.Consumer.SubscribeSubjects {
+		subscription, err := n.jsctx.PullSubscribe(subject, n.parameters.AppName, nats.BindStream(n.parameters.Stream.Name))
+		if err != nil {
+			return errors.Wrap(ErrSubscription, err.Error()+": "+subject)
+		}
+
+		n.subscriptions = append(n.subscriptions, subscription)
+	}
+
+	return nil
+}
+
+// PullMsg pulls upto batch count of messages from the stream through the pull based subscription.
+func (n *NatsJetstream) PullMsg(ctx context.Context, batch int) ([]Message, error) {
+	if n.jsctx == nil {
+		return nil, errors.Wrap(ErrNatsJetstreamAddConsumer, "Jetstream context is not setup")
+	}
+
+	msgs := []Message{}
+
+	var hasPullSubscription bool
+
+	for _, subscription := range n.subscriptions {
+		if subscription.Type() != nats.PullSubscription {
+			continue
+		}
+
+		hasPullSubscription = true
+
+		natsMsgs, err := subscription.Fetch(batch)
+		if err != nil {
+			return nil, errors.Wrap(ErrNatsMsgPull, err.Error())
+		}
+
+		for _, msg := range natsMsgs {
+			msgs = append(msgs, &NatsMsg{natsMsg: msg})
+		}
+	}
+
+	if !hasPullSubscription {
+		return nil, errors.Wrap(ErrNatsMsgPull, "no pull subscriptions to fetch from")
+	}
+
+	return msgs, nil
+}
+
+func (n *NatsJetstream) subscriptionCallback(natsMsg *nats.Msg) {
+	msg := &NatsMsg{natsMsg: natsMsg}
+
+	select {
+	case <-time.After(subscriptionCallbackTimeout):
+		_ = natsMsg.NakWithDelay(nakDelay)
+	case n.subscriberCh <- msg:
```

Comment:
why not use the builtin `ChanSubscribe` instead of doing this stuff?

Reply:
@fishnix this library returns a [`Message` interface type](https://github.com/metal-toolbox/hollow-toolbox/blob/a92687346a8c9490dcb82ebb9cf70b2751621fa3/events/events.go#L62) to a subscriber, so as to abstract interfacing with NATS or any other stream broker.

The `subscriptionCallback` packs a `*nats.Msg` into an `events.Message` which is returned to the client to keep the abstraction. Using `ChanSubscribe` would mean the client is returned a `chan *nats.Msg` directly, which pokes a hole in the abstraction. 

We can get rid of this `Message` interface abstraction if it seems overkill, although it makes the `StreamBroker` 'leaky' and swapping out the `StreamBroker` for the caller in tests is more challenging since the callers then depend on the NATS library directly to read from `chan *nats.Msg`.

Let me know what your thoughts are here.

Reply:
that makes sense, thanks 👍